### ISSUE TITLE
indentRule causing runtime error when the file structure needs to be re-ordered

### ISSUE THREAD

**Issue by christy-95** [yoheimuta/protolint#280]
When using protolint to lint some files, I realized that when the file needs to be re-ordered, runtime error will occur. For example, if the "import" statement comes before "package" statement as below:
```
syntax = "proto3";
import "template.proto"
package helloworld;
option java_package = "com.helloworld.foo"
```

It will show the following error, when you run protolint.
```
panic: runtime error: index out of range [54] with length 51

goroutine 1 [running]:
github.com/yoheimuta/protolint/internal/addon/rules.indentVisitor.validateIndent({0xc000526000, {0xc74c28, 0x2}, 0x0, 0x1, 0x0, 0xc000520060}, {{0xc0002ae28b, 0xc}, 0x543, ...}, ...)
        /home/runner/work/protolint/protolint/internal/addon/rules/indentRule.go:304 +0x505
github.com/yoheimuta/protolint/internal/addon/rules.indentVisitor.validateIndentLeading(...)
        /home/runner/work/protolint/protolint/internal/addon/rules/indentRule.go:291
github.com/yoheimuta/protolint/internal/addon/rules.indentVisitor.VisitService({0xc000526000, {0xc74c28, 0x2}, 0x0, 0x1, 0x0, 0xc000520060}, 0xc00029f5e0)
        /home/runner/work/protolint/protolint/internal/addon/rules/indentRule.go:267 +0x13c
github.com/yoheimuta/protolint/linter/visitor.extendedDisableRuleVisitor.VisitService({{0xd51128, 0xc000520090}, 0xc000528018}, 0xc00029f5e0)
        /home/runner/work/protolint/protolint/linter/visitor/extendedDisableRuleVisitor.go:150 +0xa8
github.com/yoheimuta/go-protoparser/v4/parser.(*Service).Accept(0xc00029f5e0, {0xd4f6d0, 0xc000528030})
        /home/runner/go/pkg/mod/github.com/yoheimuta/go-protoparser/v4@v4.6.0/parser/service.go:83 +0x39
github.com/yoheimuta/go-protoparser/v4/parser.(*Proto).Accept(0xc000464a50, {0xd4f6d0, 0xc000528030})
        /home/runner/go/pkg/mod/github.com/yoheimuta/go-protoparser/v4@v4.6.0/parser/proto.go:26 +0x99
github.com/yoheimuta/protolint/linter/visitor.RunVisitor({0xd51128, 0xc000520090}, 0xc000284558, {0xc76db4, 0x6})
        /home/runner/work/protolint/protolint/linter/visitor/hasExtendedVisitor.go:34 +0xca
github.com/yoheimuta/protolint/internal/addon/rules.IndentRule.Apply({{0xc74c28, 0xb2884f}, 0xc, 0x0}, 0xc00031e400)
        /home/runner/work/protolint/protolint/internal/addon/rules/indentRule.go:75 +0x117
github.com/yoheimuta/protolint/internal/linter.(*Linter).Run(0x0, 0xc000515970, {0xc000323800, 0x12, 0x0})
        /home/runner/work/protolint/protolint/internal/linter/linter.go:33 +0xf3
github.com/yoheimuta/protolint/internal/cmd/subcmds/lint.(*CmdLint).runOneFile(0xc0002a2f00, {{0xc0002ae280, 0x17}, {0xc0002ae28b, 0xc}})
        /home/runner/work/protolint/protolint/internal/cmd/subcmds/lint/cmdLint.go:138 +0x16e
github.com/yoheimuta/protolint/internal/cmd/subcmds/lint.(*CmdLint).run(0xc0002a2f00)
        /home/runner/work/protolint/protolint/internal/cmd/subcmds/lint/cmdLint.go:107 +0xb8
github.com/yoheimuta/protolint/internal/cmd/subcmds/lint.(*CmdLint).Run(0xc0002a2f00)
        /home/runner/work/protolint/protolint/internal/cmd/subcmds/lint/cmdLint.go:84 +0x4d
github.com/yoheimuta/protolint/internal/cmd.doLint({0xc0000380a0, 0x0, 0x11514a0}, {0xd32260, 0xc000006018}, {0xd32260, 0xc000006020})
        /home/runner/work/protolint/protolint/internal/cmd/cmd.go:110 +0x2df
github.com/yoheimuta/protolint/internal/cmd.doSub({0xc000038090, 0x21ce9909878, 0x0}, {0xd32260, 0xc000006018}, {0xd32260, 0xc000006020})
        /home/runner/work/protolint/protolint/internal/cmd/cmd.go:64 +0xb2
github.com/yoheimuta/protolint/internal/cmd.Do({0xc000038090, 0x60, 0x0}, {0xd32260, 0xc000006018}, {0xd32260, 0xc000006020})
        /home/runner/work/protolint/protolint/internal/cmd/cmd.go:49 +0x86
main.main()
        /home/runner/work/protolint/protolint/cmd/protolint/main.go:11 +0x6c

```

I have tested with the example code from https://github.com/protocolbuffers/protobuf/blob/main/examples/addressbook.proto (with switching around the order of "import" statement and "package" statement), I still get the similar runtime error. It seems to be the INDENT rule problem because if I disable the INDENT rule, the protolint will be able to run as expected. However, in that case, I won't be able to get any reports about the violations on indentation. Is it possible to fix it? Thank you!

**Comment by yoheimuta** [comment]
@christy-95 Thank you for reaching out.

I couldn't reproduce any issues locally.
Can you tell me how to run the protolint? Do you use a particular protolint.yaml?

```
(base) ❯ protolint version
protolint version 0.39.0(20ca4d4)

# Run it without .protolint.yaml
## Your code snippet lacks ; at the end of lines.
(base) ❯ protolint ./_example/proto/issue_280/example.proto
found "\"package\"(Token=2, Pos=_example/proto/issue_280/example.proto:3:1)" but expected [;] at /home/runner/go/pkg/mod/github.com/yoheimuta/go-protoparser/v4@v4.6.0/parser/import.go:81. Use -v for more details

(base) ❯ curl -o ./_example/proto/issue_280/addressbook.proto https://raw.githubusercontent.com/protocolbuffers/protobuf/main/examples/addressbook.proto
(base) ❯ protolint ./_example/proto/issue_280/addressbook.proto
[_example/proto/issue_280/addressbook.proto:27:1] The line length is 81, but it must be shorter than 80
[_example/proto/issue_280/addressbook.proto:37:5] EnumField name "MOBILE" should have the prefix "PHONE_TYPE"
[_example/proto/issue_280/addressbook.proto:38:5] EnumField name "HOME" should have the prefix "PHONE_TYPE"
[_example/proto/issue_280/addressbook.proto:39:5] EnumField name "WORK" should have the prefix "PHONE_TYPE"
[_example/proto/issue_280/addressbook.proto:37:5] EnumField name "MOBILE" with zero value should have the suffix "UNSPECIFIED"
```

**Comment by christy-95** [comment]
Hi, I'd tried both including and without including the .protolint.yaml file, both of them display the same error. But, one thing in common is that I used a .pre-commit-config.yaml file. 

The .pre-commit-config.yaml file. looks like this:
```
repos:
-   repo: local
    hooks:
    - id : protolint
      name: protolint
      args: [-fix]
      language: system
      entry: ./protolint.exe lint
      types: [proto]
```

The command I used to run it is:
```
 pre-commit run protolint --files ./address.proto
```

**Comment by yoheimuta** [comment]
I suspect that running it on windows may cause this error.

> It will show the following error, when you run protolint.
>> github.com/yoheimuta/protolint/internal/addon/rules.indentVisitor.VisitService({0xc000526000, {0xc74c28, 0x2}, 0x0, 0x1, 0x0, 0xc000520060}, 0xc00029f5e0)

@christy-95 How do you generate this stacktrace?
The target proto file is supposed to include the service definition, but your example/snippet lacks the service.

**Comment by christy-95** [comment]
The stack trace is from another example, the code is from https://github.com/pseudomuto/protoc-gen-doc/blob/master/examples/proto/Booking.proto

**Comment by yoheimuta** [comment]
If the target file has mixtures of line ending formats like "\n," "\r," and "\r\n," this out-of-range error makes sense.
The protolint [detects one format as a dominant line ending](https://github.com/yoheimuta/protolint/blob/4d2f2d64c047bf53f9ab3f895c041e73f9b6f2d9/internal/osutil/file.go#L63).

**Comment by christy-95** [comment]
So this means that if I want to avoid getting the out-of-range error, my file can only have either "\n" or  "\r" or "\r\n" (only one from these three line ending formats). Please correct me if I misunderstood anything.

**Comment by yoheimuta** [comment]
> my file can only have either "\n" or "\r" or "\r\n" (only one from these three line ending formats).

@christy-95 That's right.
protolint doesn't expect the mixture at the moment. 

If allowing this is in demand, I'm willing to improve it, though.

FWIW, https://github.com/pseudomuto/protoc-gen-doc/blob/master/examples/proto/Booking.proto is consistent about a line ending.

**Comment by christy-95** [comment]
In that case ( https://github.com/pseudomuto/protoc-gen-doc/blob/master/examples/proto/Booking.proto), I still get the out-of-range error, when I switch around the order import and package statements. But if I didn't switch around the order of statements, it works fine. Is it because of the OS or it's anything else?

**Comment by yoheimuta** [comment]
@christy-95 If the resulted file mixes the line ending, yes.
Once you add a new line, the error happens whether changes you made.

Can you check it using like [mixed-line-ending rule](https://gdevops.gitlab.io/tuto_git/tools/pre-commit/repos_hooks/repo_pre_commit_hooks/mixed-line-ending/mixed-line-ending.html) locally or upload the file to gist or anything while preserving a file format.

**Comment by christy-95** [comment]
I have tried using the mixed-line-ending rule, it worked! Thanks a lot!

**Comment by yoheimuta** [comment]
Good to hear that 😸

**Comment by christy-95** [comment]
Hi, I'm just wondering will the protolint be able to expect the mixture of line ending formats in near future?

**Comment by yoheimuta** [comment]
Yes! 
As I learned your use case, I'm sure supporting the mixture is worth it.

**Comment by yoheimuta** [comment]
@christy-95 Hmm... I couldn't reproduce this issue yet.
I found out that the mixture itself seemed no problem.

```
(base) ❯ echo 'syntax = "proto3";\nimport "template.proto";\r\npackage helloworld;\roption java_package = "com.helloworld.foo";' > _example/proto/issue_280/addressbook4.proto

(base) ❯ ./protolint ./_example/proto/issue_280/addressbook4.proto
[_example/proto/issue_280/addressbook4.proto:3:1] The order of Package is invalid. Check if the file is ordered in the correct manner.
[_example/proto/issue_280/addressbook4.proto:3:21] Found a possible incorrect indentation style. Inserting a new line is recommended.

(base) ❯ file ./_example/proto/issue_280/addressbook4.proto
./_example/proto/issue_280/addressbook4.proto: Perl5 module source text, ASCII text, with CRLF, CR, LF line terminators

(base) ❯ od -t c ./_example/proto/issue_280/addressbook4.proto
0000000    s   y   n   t   a   x       =       "   p   r   o   t   o   3
0000020    "   ;  \n   i   m   p   o   r   t       "   t   e   m   p   l
0000040    a   t   e   .   p   r   o   t   o   "   ;  \r  \n   p   a   c
0000060    k   a   g   e       h   e   l   l   o   w   o   r   l   d   ;
0000100   \r   o   p   t   i   o   n       j   a   v   a   _   p   a   c
0000120    k   a   g   e       =       "   c   o   m   .   h   e   l   l
0000140    o   w   o   r   l   d   .   f   o   o   "   ;  \n
0000155
```

Can you share your reproducible file with me?
Uploading the file or dumping it in both text and od format is helpful to look into it.

**Comment by christy-95** [comment]
Hi, @yoheimuta! Sorry for the late reply. Basically, this is the file [booking.txt](https://github.com/yoheimuta/protolint/files/9591491/booking.txt) I have (I'd converted into the .txt file). So, basically, it is just with one line ending format which is CRLF. After I've staged and run the pre-commit hook (with protolint), it will have the mixture (CRLF and LF) with the errors.

Thank you!

**Comment by yoheimuta** [comment]
@christy-95 Thank you for uploading the file.
I fixed it by v0.41.0. 
Can you give it a try when you have time?

**Comment by christy-95** [comment]
Hi @yoheimuta , it works well! Really appreciate it! Thank you very much! :smile:
